content/better-fors.md

```diff
@@ -333,6 +334,60 @@ doSth(arg).map { a =>
 
 This change is strictly an optimization. This allows for the compiler to get rid of the final `map` call, if the yielded value is the same as the last bound pattern. The pattern can be either a single variable binding or a tuple.
 
+This optimization should be done after type checking (e.g. around first transform). See the reasons to why it cannot be done in desugaring in [here](#previous-design-in-desugaring).
+
+We propose an approach where an attachment (`TrailingForMap`) is attached to the last `map` `Apply` node. After that, a later phase will look for `Apply` nodes with this attachment and possibly remove the `map` call.
+
+The condition for allowing to remove the last map call (for a binding `pat <- gen yield pat1`) are as follows:
+- `pat` is (syntactically) equivalent to `pat1` ($pat =_{s} pat1$)
+  
+  where
+  
+  $x =_{s} x, \text{if x is a variable reference}$
+
+  $x =_{s} (), \text{if x is a variable reference of type Unit}$
+  
+  $(x_1, ..., x_n) =_{s} (y_1, ..., y_n) \iff \forall i \in n.\; x_i =_{s} y_i$
+
+  This means that the two patterns are equivalent if they are the same variable, if they are tuples of the same variables, or if one is a variable reference of type `Unit` and the other is a `Unit` literal.
+- `pat` and `pat1` have the same types (`pat.tpe` =:= `pat1.tpe`)
+
+##### Changes discussion
+
+This adresses the problem of changing the resulting type after removing trailing `map` calls.
+
+There are two main changes compared to the previous design:
+1. Moving the implementation to the later phase, to be able to use the type information and explicitly checking that the types are the same.
+2. Allowing to remove the last `map` call if the yielded value is a `Unit` literal (and obviously the type doesn't change).
+
+The motivation for the second change is to avoid potential memory leaks in effecting loops. e.g.
+
+```scala
+//> using scala 3.3.3
+//> using lib "dev.zio::zio:2.1.5"
+
+import zio.*
+
+def loop: Task[Unit] =
+  for
+    _ <- Console.print("loop")
+    _ <- loop
+  yield ()
+
+@main
+def run =
+  val runtime = Runtime.default
+  Unsafe.unsafe { implicit unsafe =>
+    runtime.unsafe.run(loop).getOrThrowFiberFailure()
+  }
+```
+
+This kind of effect loop is pretty commonly used in Scala FP programs and often ends in `yield ()`.
+
+The problem with the desugaring of this for-comprehensions is that it leaks memory because the result of `loop` has to be mapped over with `_ => ()`, which often does nothing.
```

Comment:
Will it then be desugared to two `foreach`? as the `_` will never be used anyway.

Reply:
In this case, it should end up with just one `flatMap`. It will eliminate the last `map` call. This is also what bm4 would do in this situation.